### ISSUE TITLE
Accept iterable in each block

### ISSUE THREAD

**Issue by jankapunkt** [meteor/blaze#339]
Currently I can only pass arrays, cursors or falsey values to `#each`, otherwise I get this error:

```
 Error: {{#each}} currently only accepts arrays, cursors or falsey values.
```

However, there are widely accepted standard implementations like `Set` or `Map` that implement `iterable` and can easily be iterated using either their builtin `forEach` or using `for .. of`.

This would ease up many use-cases, where I currently have to switch between Map/Set and Array.

Involved tasks:

- Research where / how `#each` iterates and if `iterable` can be used
- Implement an iteration using `iterable`

**Comment by antoinep92** [comment]
Do you expect {{#each "string"}} to iterate over each character then ? Or should we only whitelist some classes like Map and Set ?

**Comment by antoinep92** [comment]
I can make a PR for this, but we need to decide the behavior. I see 3 options:
- accept any iterable
- accept any iterable except blacklisted types (string, etc.)
- accept anything all whitelisted types (but add Map and Set to the current whitelist)

**Comment by jankapunkt** [comment]
Great @antoinep92 !

I would rather be open here, since iterable is a well-proven interface in web-dev and I assume someone implementing it, knows what they are doing.

However, I can see, that some things just don't make sense or could cause confusion (beginners accidentally iterate over string and wonder why the see a list of characters).

I would therefore go for option two, accept any + blacklist primitives (string, number, bigint, boolean, undefined, symbol, and null), function and, of course all Objects/instances that do not implement iterable.

This is of course just my current opinion but I think it might be beneficial to let users implement their own iterable structures if they want :-D

**Comment by antoinep92** [comment]
I also think it's the better option.

Since number, undefined, etc. do not implement iterable there is no need to explicitly blacklist them. From what I've seen of the code, it seems simpler and safer to use Array.from, this incurs a minor performance penalty since it will make a shallow copy of the iterable, but for this use-case (DOM-rendering), I think this is negligible and an acceptable tradeoff.

I'll try to draft a PR this week.

**Comment by jankapunkt** [comment]
I could provide some Tests Code for some iterables

- Map
- Set
- linked list
- binary tree
- es6 Generator function

**Comment by antoinep92** [comment]
@jankapunkt I just created a PR for this. I didn't add new tests, so feel free !
Also if you can have a look at my diff and see if you spot anything, that would be appreciated.
Please note the (as far as I can tell), upstream commit 2d41716645c75c5bc2ef37f306ef87c00b982d16 is breaking the build, I don't know how this will affect the merge process :thinking:

**Comment by jankapunkt** [comment]
I will Check this out 😁

**Comment by jankapunkt** [comment]
@antoinep92 I just realized you did the PR on the meteor main repo. Is there a specific reason why not in this repo?

**Comment by antoinep92** [comment]
When I looked at the code, I realized this stuff lives in the meteor repo. I saw an issue to move to blaze, as that would indeed make more sense, but that's beyond the scope of this issue/PR

**Comment by jankapunkt** [comment]
I don't understand. The fix should be done here. Meteor only includes this repo as a git submodule. I don't think it's necessary to change any code beyond this repo.

**Comment by antoinep92** [comment]
Ok, so maybe I missed something. I'm a meteor/blaze user but first time contributor. I was looking at where the code handling {{#each}} argument is, and found that `packages/observe-sequence/observe_sequence.js` was of interest, in the *meteor* repo. This is were the argument is type-checked and an user-facing error is raised. I first tried to clone the `blaze` repo of course, but did not find high level checks there. I may be mistaken, but again this [issue](https://github.com/meteor/blaze/issues/322) makes me think it makes sense.

**Comment by filipenevola** [comment]
Yes, this package is still on Meteor repo but it could be moved to here.

We could first merge your PR and later make the move.

**Comment by StorytellerCZ** [comment]
Merge -> publish -> move

**Comment by antoinep92** [comment]
@jankapunkt The code was moved and I made a new PR in the blaze repo. I also added two very basic tests, but if you want to add a few more, feel free - at `packages/observe-sequence/observe_sequence_tests.js` and run with `test-app/ci.sh`

**Comment by jankapunkt** [comment]
I think this can be closed, since implemented via #341
